compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEGenerateColumnControlOverlay.cpp

```diff
@@ -0,0 +1,191 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree-amd-aie/Transforms/Transforms.h"
+#include "iree-amd-aie/Transforms/Utils/AMDAIEUtils.h"
+#include "iree-amd-aie/aie_runtime/Utils/ChannelGenerator.h"
+
+#define DEBUG_TYPE "iree-amdaie-generate-column-control-overlay"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+/// Initializes the channel generators for the shim tiles, excluding any
+/// channels that are already in use by existing circuit flows.
+LogicalResult initializeChannelsGenerators(
+    AMDAIE::WorkgroupOp workgroupOp, const AMDAIEDeviceModel &deviceModel,
+    const DenseSet<TileOp> &shimTileOps,
+    DenseMap<Value, ChannelGenerator> &shimTileToGeneratorMap) {
+  // Check the number of DMA channels available for the shim tile.
+  uint8_t numShimDmaChannels = deviceModel.getDmaProp<uint8_t>(
+      AMDAIETileType::SHIMNOC, AMDAIEDmaProp::NumChannels);
+  std::for_each(shimTileOps.begin(), shimTileOps.end(), [&](TileOp shimTileOp) {
+    shimTileToGeneratorMap[shimTileOp.getResult()] =
+        ChannelGenerator(numShimDmaChannels, numShimDmaChannels);
+  });
+  // Exclude those channels that are already used by a circuit flow.
+  workgroupOp->walk([&](AMDAIE::FlowOp flowOp) {
+    if (flowOp.getIsPacketFlow()) return WalkResult::advance();
+    SmallVector<AMDAIE::ChannelOp> sourceChannels;
+    for (Value source : flowOp.getSources()) {
+      if (auto channelOp =
+              dyn_cast<AMDAIE::ChannelOp>(source.getDefiningOp())) {
+        sourceChannels.push_back(channelOp);
+      }
+    }
+    for (AMDAIE::ChannelOp channelOp : sourceChannels) {
+      AMDAIE::TileOp tileOp = channelOp.getTileOp();
+      uint8_t channel = channelOp.getValue();
+      StrmSwPortType portType = channelOp.getPortType();
+      AMDAIE::DMAChannelDir direction = channelOp.getDirection();
+      if (shimTileOps.contains(tileOp) && portType == StrmSwPortType::DMA) {
+        // Assign to exclude.
+        if (direction == AMDAIE::DMAChannelDir::MM2S) {
+          shimTileToGeneratorMap[tileOp.getResult()].assignProducerDMAChannel(
+              channel);
+        } else {
```

Comment:
Use an `else if` here and add an assert in the else case to make sure we're only ever handling `MM2S` and `S2MM`.